hoodie-client/src/main/java/com/uber/hoodie/io/storage/HoodieOrcWriter.java

```diff
@@ -0,0 +1,192 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.io.storage;
+
+import com.uber.hoodie.avro.HoodieAvroWriteSupport;
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.common.model.HoodieRecordPayload;
+import com.uber.hoodie.common.util.FSUtils;
+import com.uber.hoodie.common.util.HoodieAvroUtils;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.ql.exec.vector.BytesColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.DoubleColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.LongColumnVector;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
+import org.apache.orc.CompressionKind;
+import org.apache.orc.OrcFile;
+import org.apache.orc.TypeDescription;
+import org.apache.orc.Writer;
+import org.apache.spark.TaskContext;
+
+/**
+ * HoodieOrcWriter use hive's Writer to help limit the size of underlying file. Provides
+ * a way to check if the current file can take more records with the <code>canWrite()</code>
+ */
+public class HoodieOrcWriter<T extends HoodieRecordPayload, R extends IndexedRecord>
+    implements HoodieStorageWriter<R> {
+
+  private static AtomicLong recordIndex = new AtomicLong(1);
+
+  private final Path file;
+  private final HoodieWrapperFileSystem fs;
+  private final long stripeSize;
+  private final HoodieAvroWriteSupport writeSupport;
+  private final String commitTime;
+  private final Writer writer;
+  private final TypeDescription orcSchema;
+  private final List<String> orcFieldNames;
+  private final VectorizedRowBatch orcBatch;
+
+  public HoodieOrcWriter(String commitTime, Path file, HoodieOrcConfig orcConfig,
+      TypeDescription orcSchema) throws IOException {
+    this.file = HoodieWrapperFileSystem.convertToHoodiePath(file, orcConfig.getHadoopConf());
+    this.fs = (HoodieWrapperFileSystem) this.file.getFileSystem(registerFileSystem(file, orcConfig.getHadoopConf()));
+    this.writeSupport = orcConfig.getWriteSupport();
+    this.commitTime = commitTime;
+    this.stripeSize = orcConfig.getStripeSize();
+
+    this.orcSchema = orcSchema;
+    // hoodie meta field
+    this.orcSchema.addField("_hoodie_commit_time", TypeDescription.createString())
```

Comment:
HoodieAvroUtils has a method to add metadata to the schema, if needed add a method there and use it here.

Reply:
> HoodieAvroUtils has a method to add metadata to the schema, if needed add a method there and use it here.

very good , thanks.